### PR TITLE
Improve data loading multi thread performance

### PR DESCRIPTION
**Pull request by john-sp**
This PR combines two improvements to the data loading pipeline:
1. Moves mutex protection to the file source level in the tar chunk reader to reduce contention and improve thread safety when reading from multiple files concurrently.
2. Adds batch prefetching in the Python training loop to overlap data preparation with computation, improving training throughput. (tested only a few seconds faster per 5k steps)
Additionally, the C++ chunk source loader was modernized to use file descriptors with pread() for thread-safe reading, replaced FILE* usage, added proper mutex protection for shared data structures, and used shared_ptr for better lifetime management across threads. These changes collectively improve the performance and reliability of the data loading system.

### CODE REVIEW

File: csrc/loader/chunk_source/tar_chunk_source.cc [hunk 1]
```diff
@@ -46,13 +51,33 @@ uint64_t ParseOctal(const std::array<uint8_t, 12>& octal) {
   return value;
 }
 
-std::optional<std::string> ReadGzipPrefix(FILE* file, long int offset,
-                                          long int size, size_t max_bytes) {
-  if (max_bytes == 0) return std::string();
-
-  if (fseek(file, offset, SEEK_SET) != 0) {
-    return std::nullopt;
+bool ReadExact(int fd, off_t offset, void* buffer, size_t size) {
```
Comment (mooskagh) [inline csrc/loader/chunk_source/tar_chunk_source.cc:52]: This function takes offset but doesn't use it?

File: csrc/loader/chunk_source/tar_chunk_source.cc [hunk 2]
```diff
@@ -46,13 +51,33 @@ uint64_t ParseOctal(const std::array<uint8_t, 12>& octal) {
...
+  if (offset < 0) return false;
+  if (size > static_cast<size_t>(std::numeric_limits<ssize_t>::max())) {
```
Comment (mooskagh) [inline csrc/loader/chunk_source/tar_chunk_source.cc]: Is this check needed for something? Looks quite useless.

File: csrc/loader/chunk_source/tar_chunk_source.cc (hunk 1 above)
Comment (mooskagh) [reply csrc/loader/chunk_source/tar_chunk_source.cc:52]: Ah sorry, I accidentally included comma in the search.

File: csrc/loader/chunk_source/tar_chunk_source.cc [hunk 3]
```diff
@@ -46,13 +51,33 @@ uint64_t ParseOctal(const std::array<uint8_t, 12>& octal) {
...
+    return false;
   }
+  char* out = static_cast<char*>(buffer);
+  size_t read_total = 0;
+  while (read_total < size) {
+    if (offset > std::numeric_limits<off_t>::max() -
```
Comment (mooskagh) [inline csrc/loader/chunk_source/tar_chunk_source.cc]: This also appears quite useless..

File: csrc/loader/chunk_source/tar_chunk_source.cc [hunk 4]
```diff
@@ -46,13 +51,33 @@ uint64_t ParseOctal(const std::array<uint8_t, 12>& octal) {
...
+                     static_cast<off_t>(read_total)) {
+      return false;
+    }
+    const ssize_t read_now =
+        pread(fd, out + read_total, size - read_total, offset + read_total);
+    if (read_now < 0) {
+      if (errno == EINTR) continue;
```
Comment (mooskagh) [inline csrc/loader/chunk_source/tar_chunk_source.cc]: not needed. And then we can collapse condition to `read_now <= 0`

File: csrc/loader/chunk_source/tar_chunk_source.cc [hunk 5]
```diff
@@ -46,13 +51,33 @@ uint64_t ParseOctal(const std::array<uint8_t, 12>& octal) {
...
 }
 
-std::optional<std::string> ReadGzipPrefix(FILE* file, long int offset,
-                                          long int size, size_t max_bytes) {
-  if (max_bytes == 0) return std::string();
-
-  if (fseek(file, offset, SEEK_SET) != 0) {
-    return std::nullopt;
+bool ReadExact(int fd, off_t offset, void* buffer, size_t size) {
+  if (offset < 0) return false;
```
Comment (mooskagh) [inline csrc/loader/chunk_source/tar_chunk_source.cc]: That's also not needed. We can at most use assert()..

File: csrc/loader/chunk_source/tar_chunk_source.cc [hunk 6]
```diff
@@ -113,29 +138,44 @@ std::optional<std::string> ReadGzipPrefix(FILE* file, long int offset,
...
   Index();
 }
 
-TarChunkSource::~TarChunkSource() {
-  if (file_) fclose(file_);
+TarChunkSource::~TarChunkSource() { Close(); }
+
+void TarChunkSource::Close() {
+  if (fd_ >= 0) {
+    if (close(fd_) != 0) {
```
Comment (mooskagh) [inline csrc/loader/chunk_source/tar_chunk_source.cc]: I'd combine two `if`s into one. (and assign -1 unconditionally)

File: csrc/loader/chunk_source/tar_chunk_source.cc [hunk 7]
```diff
@@ -151,19 +191,23 @@ void TarChunkSource::Index() {
...
-    const long int offset = ftell(file_);
-    const long int size = ParseOctal(header.size);
-    const long int new_offset = offset + (size + 511) / 512 * 512;
-    fseek(file_, new_offset, SEEK_SET);
-    if (new_offset != ftell(file_)) {
+    const off_t file_offset = offset;
+    const size_t size = ParseOctal(header.size);
+    const uint64_t padded_size =
+        ((static_cast<uint64_t>(size) + 511) / 512) * 512;
+    if (padded_size > static_cast<uint64_t>(std::numeric_limits<off_t>::max()) ||
```
Comment (mooskagh) [inline csrc/loader/chunk_source/tar_chunk_source.cc]: These check for overflows are not needed either. (also withoug ftell we won't know it's truncated before we read, so this LOG(WARNING) has to be removed too)

File: src/lczero_training/training/training.py
Comment (mooskagh) [inline src/lczero_training/training/training.py:1]: What is the intention here? It doesn't appear the changes in this file are useful enough to justify adding complexity?

**Review by mooskagh — CHANGES_REQUESTED** [review]

File: src/lczero_training/training/training.py
Comment (john-sp) [reply src/lczero_training/training/training.py:1]: This is loading the next batch to the GPU while the current batch is running, saved me roughly 30 seconds per 5k steps (40 minutes).

**Review by mooskagh — APPROVED** [review]
csrc/loader/stages/shuffling_chunk_pool.cc is a mess of course (even before this change)... I hope to clean it up one day.

File: src/lczero_training/training/training.py
Comment (mooskagh) [reply src/lczero_training/training/training.py:1]: But we don't have 30 seconds per step? Our training step is like 1/3 of a second?

**Review by mooskagh — CHANGES_REQUESTED** [review]

File: src/lczero_training/training/training.py
Comment (john-sp) [reply src/lczero_training/training/training.py:1]: It was saving 0.006 seconds per step, that is most like within the training noise. Reverted this file

**Review by mooskagh — APPROVED** [review]
